### PR TITLE
Enable nullability in BasicDesignerLoader and CodeDomDesignerLoader

### PR DESCRIPTION
**Pull request by halgab**
## Proposed changes

- Enable nullability in `BasicDesignerLoader`
- Enable nullability in `CodeDomDesignerLoader`

This includes some code cleanup, and consolidation between the two classes

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9779)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs [hunk 1]
```diff
@@ -40,39 +38,19 @@ public bool CanExtend(object o)
...
-        private IComponent GetBaseComponent(IComponent c)
+        private IComponent? GetBaseComponent(IComponent c)
         {
-            IComponent baseComponent = null;
-
-            if (c is null)
-            {
-                return null;
-            }
+            _host ??= c.Site?.GetService<IDesignerHost>();
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:51]: In the old code, we checked if `c is null`. Do we never get null for c?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs [hunk 2]
```diff
@@ -88,22 +66,11 @@ private IComponent GetBaseComponent(IComponent c)
         [HelpKeyword("Designer_GenerateMember")]
         public static bool GetGenerateMember(IComponent comp)
         {
-            ISite site = comp.Site;
-
-            if (site is null)
+            if (comp.Site.TryGetService(out IDictionaryService? dictionary))
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:69]: I think we could get a null reference here now. We used to return `true` if `comp.Site is null`.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs [hunk 3]
```diff
@@ -125,31 +92,20 @@ public static bool GetGenerateMember(IComponent comp)
         [HelpKeyword("Designer_Modifiers")]
         public static MemberAttributes GetModifiers(IComponent comp)
         {
-            ISite site = comp.Site;
-
-            if (site is not null)
+            if (comp.Site.TryGetService(out IDictionaryService? dictionary))
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:95]: Same here, seems like we may get a null reference for `comp.Site`

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs [hunk 4]
```diff
@@ -208,16 +165,10 @@ public static void SetGenerateMember(IComponent comp, bool generate)
         /// </summary>
         public static void SetModifiers(IComponent comp, MemberAttributes modifiers)
         {
-            ISite site = comp.Site;
-
-            if (site is null)
+            if (comp.Site.TryGetService(out IDictionaryService? dictionary))
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:168]: Same here it looks like `comp.Site` could be null based off of old code.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.cs [hunk 5]
```diff
@@ -269,15 +265,15 @@ private void EnsureDocument(IDesignerSerializationManager manager)
...
+                                    Type? serializerType = LoaderHost.GetType(ra.SerializerTypeName!);
 
                                     if (serializerType is not null)
                                     {
                                         foundAttribute = true;
 
                                         if (firstClass)
                                         {
-                                            _rootSerializer = (CodeDomSerializer)Activator.CreateInstance(serializerType, BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.CreateInstance, null, null, null);
+                                            _rootSerializer = (CodeDomSerializer?)Activator.CreateInstance(serializerType, BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.CreateInstance, null, null, null);
```
Comment (lonitra) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.cs:276]: nit: could you put variable names in for the nulls

**Review by lonitra — COMMENTED** [review]
Awesome work here!

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs (hunk 1 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:51]: This private method is only called on line 32, on which `c` cannot be null per the pattern matching above on line 26. So I figured we could safely spare ourselves this null check

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs (hunk 2 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:69]: Actually [TryGetService](https://github.com/dotnet/winforms/blob/a01f11db140935fe576d1caa1b916697b81d9031/src/System.Windows.Forms.Primitives/src/System/ServiceExtensions.cs#L40) checks `service` for null. That's why I tend not to use it much, but here it makes sense.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs (hunk 3 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:95]: see https://github.com/dotnet/winforms/pull/9779/files#r1306122388

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs (hunk 4 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:168]: see https://github.com/dotnet/winforms/pull/9779/files#r1306122388

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.cs (hunk 5 above)
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.cs]: done

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/BasicDesignerLoader.cs [hunk 6]
```diff
@@ -168,21 +166,13 @@ public override void BeginLoad(IDesignerLoaderHost host)
 
             // Add our services.  We do IDesignerSerializationManager separate because
             // it is not something the user can replace.
-            DesignSurfaceServiceContainer dsc = GetService(typeof(DesignSurfaceServiceContainer)) as DesignSurfaceServiceContainer;
-
-            if (dsc is not null)
+            if (TryGetService(out DesignSurfaceServiceContainer? dsc))
```
[the remark this answers is not in the document]
Comment (halgab) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/BasicDesignerLoader.cs:169]: I tried to address as many places as I could find. I may have miss some.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs (hunk 2 above)
Comment (lonitra) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomDesignerLoader.ModifiersExtenderProvider.cs:69]: Ah I see, thanks!

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/BasicDesignerLoader.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lonitra) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/BasicDesignerLoader.cs:169]: Thank you! Appreciate it.

**Review by lonitra — DISMISSED** [review]
🚀

File: src/System.Windows.Forms.Primitives/src/System/ComponentModel/TypeDescriptorHelper.cs [hunk 7]
```diff
@@ -13,4 +13,19 @@ public static bool TryGetAttribute
         attribute = TypeDescriptor.GetAttributes(component)[typeof(T)] as T;
         return attribute is not null;
     }
+
+    public static bool TryGetPropertyValue<T>(
+        object component,
+        string name,
+        out T? attribute)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Primitives/src/System/ComponentModel/TypeDescriptorHelper.cs:20]: `attribute` is a confusing name for this argument. Consider `value`.

Comment (halgab) [reply src/System.Windows.Forms.Primitives/src/System/ComponentModel/TypeDescriptorHelper.cs]: Sorry, looks like a bad copy/paste from the method above. I've just fixed it

**Review by Tanya-Solyanik — APPROVED** [review]
👍

**Pull request merged**